pkg/storagev2/migration/migration_test.go

```diff
@@ -0,0 +1,304 @@
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "not missing",
+			input: migration.Steps{
+				1: func(s storage.Store) error {
+					return s.Put(objT1)
+				},
+				2: func(s storage.Store) error {
+					return s.Put(objT2)
+				},
+				3: func(s storage.Store) error {
+					return s.Put(objT3)
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "desc order versions",
+			input: migration.Steps{
+				3: func(s storage.Store) error {
+					return s.Put(objT1)
+				},
+				2: func(s storage.Store) error {
+					return s.Put(objT2)
+				},
+				1: func(s storage.Store) error {
+					return s.Put(objT3)
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "desc order version missing",
+			input: migration.Steps{
+				4: func(s storage.Store) error {
+					return s.Put(objT1)
+				},
+				2: func(s storage.Store) error {
+					return s.Put(objT2)
+				},
+				1: func(s storage.Store) error {
+					return s.Put(objT3)
+				},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			if err := migration.ValidateVersions(tt.input); (err != nil) != tt.wantErr {
+				t.Errorf("ValidateVersions() unexpected error: %v, wantErr : %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestMigrate(t *testing.T) {
+	t.Parallel()
+	objT1 := &obj1{Id: "aaa", SomeInt: 1}
+	objT2 := &obj1{Id: "bbb", SomeInt: 2}
+	objT3 := &obj1{Id: "ccc", SomeInt: 3}
+
+	t.Run("migration: 0 to 3", func(t *testing.T) {
+		t.Parallel()
+
+		steps := migration.Steps{
+			1: func(s storage.Store) error {
+				return s.Put(objT1)
+			},
+			2: func(s storage.Store) error {
+				return s.Put(objT2)
+			},
+			3: func(s storage.Store) error {
+				return s.Put(objT3)
+			},
+		}
+
+		s := inmemstore.New()
+
+		if err := migration.Migrate(s, steps); err != nil {
+			t.Errorf("Migrate() unexpected error: %v", err)
+		}
+
+		newVersion, err := migration.Version(s)
+		if err != nil {
+			t.Errorf("Version() unexpected error: %v", err)
+		}
+		if newVersion != 3 {
+			t.Errorf("new version = %v must be 3", newVersion)
+		}
+
+		assertObjectExists(t, s, objT1, objT2, objT3)
+	})
+
+	t.Run("migration: 5 to 8", func(t *testing.T) {
+		t.Parallel()
+
+		steps := migration.Steps{
+			8: func(s storage.Store) error {
+				return s.Put(objT1)
+			},
+			7: func(s storage.Store) error {
+				return s.Put(objT2)
+			},
+			6: func(s storage.Store) error {
+				return s.Put(objT3)
+			},
+		}
+
+		s := inmemstore.New()
+
+		err := migration.SetVersion(s, 5)
+		if err != nil {
+			t.Errorf("SetVersion() unexpected error: %v", err)
+		}
+
+		if err := migration.Migrate(s, steps); err != nil {
+			t.Errorf("Migrate() unexpected error: %v", err)
+		}
+
+		newVersion, err := migration.Version(s)
+		if err != nil {
+			t.Errorf("Version() unexpected error: %v", err)
+		}
+		if newVersion != 8 {
+			t.Errorf("new version = %v must be 8", newVersion)
+		}
+
+		assertObjectExists(t, s, objT1, objT2, objT3)
+	})
```

Comment:
One more test case to add here is failure in one of the step function. If that happens we need to ensure that the version was correctly set to the last successful migration step.

Reply:
Done